feat(client): add whereUnique support for relations with compound unique constraints

addy-2709genius opened the pull request:
## Description

Implements `whereUnique` option for relation selections when filtering by compound unique constraints, as requested in #28376.

## Problem

When selecting a relation with a compound unique constraint (e.g., `@@unique([creatorId, formId])`), using `where` returns an array even though the constraint guarantees at most 1 result.

## Solution

Adds `whereUnique` option that:
- Accepts partial compound unique constraint fields
- Returns `T | null` instead of `T[]` in types
- Converts to proper `where` clause with compound unique constraint at runtime
- Validates that `where` and `whereUnique` cannot be used together

## Example

// Before: returns FormResponse[]
const result = await prisma.form.findMany({
  select: {
    responses: {
      where: { creatorId: "..." },
    },
  },
});

// After: returns FormResponse | null
const result = await prisma.form.findMany({
  select: {
    responses: {
      whereUnique: { creatorId: "..." },
    },
  },
});## Changes

-  TypeScript generator: Added `whereUnique` to relation selection types
-  Runtime: Convert `whereUnique` to `where` with compound unique constraints
-  Tests: Comprehensive test suite added

## Note

Result processing to convert arrays to single items when `whereUnique` is used is still needed and will be added in a follow-up.

Closes #28376

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=28704) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=28704) before we can accept your contribution.<br/><hr/>**aaditya raj soni** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=28704) it.</sub>

addy-2709genius committed:
feat(client): add whereUnique support for relations with compound unique constraints

Implements whereUnique option for relation selections when the target model
has compound unique constraints. This allows filtering relations to return
a single result (or null) instead of an array, providing better type safety
when querying relations with compound unique constraints.

Changes:
- Add whereUnique field to relation selection input types in TypeScript generator
- Add runtime support for converting whereUnique to proper where clauses
- Add comprehensive test suite covering all scenarios
- Validate that where and whereUnique cannot be used together

Note: Result processing to convert arrays to single items when whereUnique is used
is still needed and will be added in a follow-up commit.

Closes #28376

addy-2709genius committed:
Merge branch 'main' into feat/client-where-unique-compound-constraints

addy-2709genius committed:
Merge branch 'main' into feat/client-where-unique-compound-constraints
